src/handlers.cpp

```diff
@@ -1579,28 +1632,129 @@ void ImpuRegDataTask::on_sar_response(Diameter::Message& rsp)
         TRC_DEBUG("Getting User-Data from SAA for cache");
         saa.user_data(_xml);
         put_in_cache();
+        caching = true;
       }
-      send_reply();
       break;
     case 5001:
     {
       TRC_INFO("Server-Assignment answer with result code %d - reject", result_code);
       SAS::Event event(this->trail(), SASEvent::REG_DATA_HSS_FAIL, 0);
       SAS::report_event(event);
-      send_http_reply(HTTP_NOT_FOUND);
+      _http_rc = HTTP_NOT_FOUND;
     }
     break;
     default:
       TRC_INFO("Server-Assignment answer with result code %d - reject", result_code);
       SAS::Event event(this->trail(), SASEvent::REG_DATA_HSS_FAIL, 0);
       SAS::report_event(event);
-      send_http_reply(HTTP_SERVER_ERROR);
+      _http_rc = HTTP_SERVER_ERROR;
       break;
   }
-  delete this;
+
+  // If we're not caching the registration data, check for deregistration and
+  // delete the task now
+  if (!caching)
```

Comment:
I was more observing that maybe we should pull the `put_in_cache` at line 1634 to here, so that we have a single block of code responsible for managing the cache, rather than doing some cache management above and then remembering whether we did it and doing the rest of the cache management down here.